File: 백준/Silver/7007. BSP Trees/BSP Trees.cc
#include <iostream>
#include <stdio.h>
using namespace std;

int main() {
	int nP;
	scanf("%d", &nP);
	int xP[26], zP[26];
	for(int i = 0; i < nP; i++){
		int num;
		scanf("%d", &num);
		for(int j = 0; j < num; j++){
			scanf("%d%d", &xP[i], &zP[i]);
		}
	}
	int nL;
	scanf("%d", &nL);
	int A[26], B[26], C[26];
	for(int i = 0; i < nL; i++){
		int x1, z1, x2, z2;
		scanf("%d%d%d%d", &x1, &z1, &x2, &z2);
		A[i] = z1-z2, B[i] = x2-x1, C[i] = x1*z2-z1*x2;
		if(B[i] < 0){
			A[i] = -A[i], B[i] = -B[i], C[i] = -C[i];
		}
	}
	int order[26];
	bool isDiv[27] = {0};
	isDiv[0] = 1, isDiv[nP] = 1;
	for(int i = 0; i < nP; i++) order[i] = i;
	for(int i = 0; i < nL; i++){
		int pnc = 0;
		while(pnc < nP){
			int next;
			for(int j = pnc+1; ; j++){
				if(isDiv[j]) {
					next = j;
					break;
				}
			}
			if(next == pnc+1){
				//区间长度为1，不用再分
				pnc = next;
				continue;
			}
			int xianNo = 0;
			int tempHou[26];
			int houNo = 0;
			for(int j = 0; j < next-pnc; j++){
				if(A[i] * xP[order[pnc+j]] + B[i] * zP[order[pnc+j]]+ C[i] > 0){
					order[pnc+xianNo] = order[pnc+j];
					xianNo ++;
				}
				else{
					tempHou[houNo] = order[pnc+j];
					houNo ++;
				}
			}
			isDiv[pnc+xianNo] = 1;
			for(int j = 0; j < houNo; j++){
				order[pnc+xianNo+j] = tempHou[j];
			}
			pnc = next;
		}
	}
	for(int i = 0; i < nP; i++){
		printf("%c", order[i]+'A');
	}
	printf("\n");
	return 0;
}

File: 백준/Silver/7007. BSP Trees/BSP Trees.py
n = int(input())
xP, zP = [0] * n, [0] * n
for i in range(n):
    *_, x, z = map(int, input().split())
    xP[i] = x
    zP[i] = z

p = int(input())
A, B, C = [0] * p, [0] * p, [0] * p

for i in range(p):
    x1, z1, x2, z2 = map(int, input().split())
    A[i] = z1 - z2
    B[i] = x2 - x1
    C[i] = x1 * z2 - z1 * x2
    if B[i] < 0: A[i], B[i], C[i] = -A[i], -B[i], -C[i]

order = list(range(n))
isDiv = bytearray(n + 1)
isDiv[0], isDiv[n] = True, True

for i in range(p):
    pnc = 0
    while pnc < n:
        next_div = pnc + 1
        while not isDiv[next_div]:
            next_div += 1
        
        if next_div == pnc + 1:
            pnc = next_div
            continue
        
        xianNo, houNo = 0, 0
        tempHou = [0] * n
        
        for j in range(next_div - pnc):
            if A[i] * xP[order[pnc + j]] + B[i] * zP[order[pnc + j]] + C[i] > 0:
                order[pnc + xianNo] = order[pnc + j]
                xianNo += 1
            else:
                tempHou[houNo] = order[pnc + j]
                houNo += 1
        
        isDiv[pnc + xianNo] = True
        
        for j in range(houNo):
            order[pnc + xianNo + j] = tempHou[j]
        
        pnc = next_div

print("".join(chr(order[i] + ord('A')) for i in range(n)))